### PR TITLE
`violin`: honor `bandwidth` attribute even if values in a trace are identical

### LINKED ISSUE #2779: Improve bandwidth calculation for single-value violins
From @alexcjohnson in https://github.com/plotly/plotly.js/pull/2775

> But if we think this is a real use case we should use other violins on the same subplot to pick a bandwidth that's reasonable for the whole collection, like we do with [histograms](https://github.com/plotly/plotly.js/blob/master/src/traces/histogram/calc.js#L350)

### LINKED ISSUE #3622: `violin` trace: why an area is calculated if all values are 0?
Please see my question in [stackoverflow](https://stackoverflow.com/questions/55094041/plotly-js-why-an-area-is-calculated-if-all-values-are-0?noredirect=1#comment96932560_55094041)

I have an array of values from 0-10 with and 20 groups in each. For some groups or categories all values are 0. When I use violin plot, there is an area calculated which is below 0, but I want to get thedifferent that I got from Seurat package VlnPlot. Is there a way to overcome that problem? Here is a part of java script code:

```
var my_violin = [];
var vln_layout = {};
var my_violin =[
{
    type:"violin",
    x:xvalues,
    y:my_categories,
    points:"all",
    pointpos:0,
    color:"black",
    width:0.75,
    marker:{color:"black",size:4},
    jitter:1,
    span:0,
    transforms:[
    {
        type:"groupby",
        groups:my_categories,
        styles:my_styles
    }
    ]
}
]

var vln_layout = {title: "my title",    xaxis:{showline: false,showgrid: false,zeroline: false,
    categoryorder: 'category ascending'}, yaxis: {showline: false,showgrid: false,zeroline:false},
    showlegend:true,legend:{traceorder:'normal',categoryorder:"ascending"}};

Plotly.newPlot('myDiv2', my_violin, vln_layout, {showSendToCloud: false});
```

### PR DESCRIPTION
**Pull request by antoinerg**
Following botched PR https://github.com/plotly/plotly.js/pull/3625, this PR closes #3622 by honoring `trace.bandwidth` even if all the values are identical.

When no bandwidth is specified and all values are identical, it draws the violin as a line. Therefore, this PR also closes #2779!

### CODE REVIEW

**Comment by antoinerg** [comment]
At the moment, the default bandwidth for a violin with zero standard deviation (ie. all values are identical) is 1. Therefore, you will get a significant violin area as shown in this [comment](https://github.com/plotly/plotly.js/issues/3622#issuecomment-471691858) which is not what other libraries do (like Seaborn for example). For people who don't know the details of violin plot (like me yesterday), this looks weird/wrong.

@alexcjohnson do you think we should greatly reduce the bandwidth when all values are identical so the violin collapses?

**Comment by alexcjohnson** [comment]
Right, I think the *default* behavior with all identical values should be zero width, as you had in #3625. But if someone provides an explicit `bandwidth`, we should honor that as you do here.

**Comment by antoinerg** [comment]
> Right, I think the default behavior with all identical values should be zero width, as you had in #3625. But if someone provides an explicit bandwidth, we should honor that as you do here.

With commit 5acb518 we have a mock showcasing both situations (identical values with and without explicit bandwidth). I had to explicitly set the bandwidth to 1 on an existing mock to preserve baselines intact.

File: test/image/mocks/groups-over-matching-axes.json [hunk 1]
```diff
@@ -86,6 +86,9 @@
     }
   ],
   "layout": {
+    "template": {
+      "data": {"violin": [{"bandwidth": 1}]}
```
Comment (alexcjohnson) [inline test/image/mocks/groups-over-matching-axes.json]: Nice use of `template` 🏆

**Comment by alexcjohnson** [comment]
💃 
This seems like more than a bugfix, so merge if you're ready for 1.46 features.

**Comment by etpinard** [comment]
@antoinerg looks like this PR resolves https://github.com/plotly/plotly.js/issues/2779 as well, correct?

File: test/image/mocks/groups-over-matching-axes.json (hunk 1 above)
Comment (etpinard) [reply test/image/mocks/groups-over-matching-axes.json]: Hmm. I think I would prefer updating the baselines instead of tweaking the JSON mocks.

**Comment by antoinerg** [comment]
> @antoinerg looks like this PR resolves #2779 as well, correct?

Yes it does! I updated my first comment to reflect that. Nice catch :eyes:

File: test/image/mocks/groups-over-matching-axes.json (hunk 1 above)
Comment (antoinerg) [reply test/image/mocks/groups-over-matching-axes.json]: Done in 4698df0

Comment (alexcjohnson) [reply test/image/mocks/groups-over-matching-axes.json]: Not a big deal, but to me the original baselines are a lot easier to look at and understand what's going on. But of course either way it'll serve the function of catching unintended changes.

**Comment by etpinard** [comment]
good to merge :dancer:

**Comment by dellshape** [comment]
Hi, updating here, passed by here because I got the same problem just now. But turns out now you can solve this by setting spanmode to "hard", therefore the "span goes from the sample's minimum to its maximum value."

**Pull request merged**